recipes/ncbi-cxx-toolkit-public/26.0/conanfile.py

```diff
@@ -0,0 +1,1587 @@
+"UV":           ("libuv/1.42.0",        "libuv/1.42.0",     "libuv/1.42.0"),
+"Z":            ("zlib/1.2.11",         "zlib/1.2.11",      "zlib/1.2.11"),
+    }
+    Conan_package_options = {
+        "libnghttp2": {
+            "with_app": False,
+            "with_hpack": False
+        },
+        "grpc": {
+            "cpp_plugin": True,
+            "csharp_plugin": False,
+            "node_plugin": False,
+            "objective_c_plugin": False,
+            "php_plugin": False,
+            "python_plugin": False,
+            "ruby_plugin": False
+        }
+    }
+    generators = "cmake"
+
+#----------------------------------------------------------------------------
+    def set_version(self):
+        if self.version == None:
+            self.version = "26.0.1"
+
+    def __del__(self):
+        if os.path.isdir(self.tk_tmp_tree):
+            print("Just a moment...")
+            self._remove_tmp_tree(self.tk_tmp_tree)
+
+    def _remove_tmp_tree(self, path):
+        if os.path.isdir(path):
+            os.chdir(os.path.join(self.tk_tmp_tree, ".."))
+            try:
+                shutil.rmtree(path, onerror=_remove_readonly)
+            except:
+                print("Cannot remove directory: ", path)
+                print("Please remove it manually")
+
+#----------------------------------------------------------------------------
+    def _get_RequiresMapKeys(self):
+        return self.NCBI_to_Conan_requires.keys()
+
+#----------------------------------------------------------------------------
+    def _translate_ReqKey(self, key):
+        index = 0
+        if self.settings.os == "Linux":
+            index = 0
+        elif self.settings.os == "Macos":
+            index = 1
+        elif self.settings.os == "Windows":
+            index = 2
+        if key in self.NCBI_to_Conan_requires.keys() and self.NCBI_to_Conan_requires[key][index]:
+            return self.NCBI_to_Conan_requires[key][index]
+        return None
+
+#----------------------------------------------------------------------------
+    def _get_Source(self):
+        self.tk_tmp_tree = tempfile.mkdtemp(dir=os.getcwd())
+        src = os.path.normpath(os.path.join(self.recipe_folder, "..", "source", "src"))
+        src_found = False;
+        if os.path.isdir(src):
+            self.tk_src_tree = src
+            src_found = True;
+        else:
+            print("getting Toolkit sources...")
+            tk_url = self.conan_data["sources"][self.version]["url"] if "url" in self.conan_data["sources"][self.version].keys() else ""
+            tk_git = self.conan_data["sources"][self.version]["git"] if "git" in self.conan_data["sources"][self.version].keys() else ""
+            tk_branch = self.conan_data["sources"][self.version]["branch"] if "git" in self.conan_data["sources"][self.version].keys() else "master"
+            tk_svn = self.conan_data["sources"][self.version]["svn"] if "svn" in self.conan_data["sources"][self.version].keys() else ""
```

Comment:
All sources should be declared in the `conandata.yml` file. From my point of view, it is deterministic, so there is no need to check any `if`.

Also, those sources should use checksums and point to a concrete commit (we need reproducibility of the packages). So don't use `master`.

Anyway, aren't these extra sources already available in the `release-26.0.1` you are already downloading? If so, a standard implementation of the `source` method should be enough.

Reply:
The "master" recipe is in github (https://github.com/ncbi/ncbi-cxx-toolkit-conan), and its scope is bigger than that of Conan center. My intention was to make minimal changes when submitting it to Center. If some code is not required, so what?
Well, if I have to remove unused code, I will.

Reply:
In my opinion, there is nothing to discuss here. "My baby" is always the best. "Your baby" is CCI, To me, even Github is not a master. "My baby" is what sits on NCBI intranet server. It is the way it is, nothing will change that.
Sync recipes? Well... I do not know, Why? It is release, it works, presumably. It is stable.
New developments will bring new problems. and there will be new recipe to solve them.

Reply:
Sure, what I was trying to say is that we would like to find a solution that fits more than 1,000 _babies_ that are right now in CCI... and we would love to hear more about their point of view about this problem so we can consider it in order to design a better solution.

Reply:
I apologize for being impolite, but do you people know anything about "backward compatibility"??
Just few hours ago somebody has "improved" libtiff package, and it now requires zstd/1.5.1
And I now have conflicts to resolve:

    ERROR: Conflict in libmysqlclient/8.0.25:
        'libmysqlclient/8.0.25' requires 'zstd/1.5.0' while 'libtiff/4.3.0' requires 'zstd/1.5.1'.
        To fix this conflict you need to override the package 'zstd' in your root package.

Well, maybe there is a way to resolve them automatically, which I do not know?
Let us say, you accept this recipe in to your repo. Should I then test it every day to make sure it still works?? That somebody did not break it somehow by "improving" something else?

Reply:
This is a problem that will happen from time to time in you consume the recipe from the HEAD of the repository (or latest recipe revision). But Conan provides two features to avoid this kind of breakage for consumers: recipe revisions and lockfiles. You can find a brief summary about them here: https://github.com/conan-io/conan-center-index/blob/master/docs/consuming_recipes.md

About the recipe itself and validation of all the graph of dependencies in ConanCenter, of course it is a problem, and it is impossible to guarantee that all the recipes are compatible at a given commit. We try to do our best, and you will see many PRs upgrading the requirements of different packages so the HEAD of the repository keeps working. We have plans to automate the creation of those PRs: whenever a new version is merged, open PRs to dependents upgrading that requirement (but we need to measure first how many PRs and how to track invalid versions for each of the consumers so we don't open again and again the same PR).

We would really like to solve this problem (a problem that is inherent to dependency management), but probably we can only try to alleviate the pain. Even if the test is conceptually simple, it is very challenging: checking if a new version of a requirement works for a consumer is a process that can take several hours (building all the configurations of some libraries takes hours), and then we need to test all the consumers of that consumer... basically, for every new recipe revision (every PR) we would need to rebuild many consumer packages. I'm afraid it is something not possible with a finite number of resources.

This is how I see the problem, maybe I'm wrong and there is some smart solution that we can use. We are really eager to hear about it, as we are about other issues that we haven't managed to figure out a possible solution to.

I really appreciate you telling about your pain, this is how we can prioritize some features over others. Besides reviewing and approving PRs, we keep working on the CI library.

Reply:
So, "recipe revision" is the magic. So far I thought it was for smaller things. I thought backward compatibility should be handled by package version.